File: tramaj-hs/src/Tramaj/Node.hs
-- | The evaluated document representation -- Tramaj's output domain and its
-- portable interchange boundary. 'nodeToJson'\/'nodeFromJson' implement the
-- normative representation specified in @../specs/node-json.md@; that
-- document, not this module, is the contract other implementations and hosts
-- are held to.
--
-- Deliberately independent of both the expression AST ("Tramaj.Ast") and of
-- any particular render target: a host folds a 'Node' into HTML, a UI tree,
-- YAML, HCL or anything else on its own terms.
module Tramaj.Node
  ( Node (..)
  , NodeAttribute (..)
  , Annotations
  , noAnnotations
  , nodeToJson
  , nodeFromJson
  , nodeAttributeToJson
  , mapActions
  ) where

import Data.Aeson (Value (..), object, (.=))
import qualified Data.Aeson.Key as Key
import qualified Data.Aeson.KeyMap as KeyMap
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Text (Text)
import qualified Data.Vector as V

-- | Arbitrary host\/tooling metadata hung off any node. The core language
-- assigns no meaning to any key; unknown annotations must not affect
-- semantics, and every node-to-node transformation must carry them through
-- unchanged. This is where a future type\/domain\/constraint pass puts what
-- it derives, without the 'Node' constructors having to change.
type Annotations = Map Text Value

noAnnotations :: Annotations
noAnnotations = Map.empty

-- | Three constructors, with enough expressivity inside them to carry
-- scalars losslessly -- see @../specs/decisions.md@ #5.
--
-- 'NText' holds a 'Value', not a 'Text': @.p($ctx.count)@ with @count = 3@
-- keeps the number @3@ rather than stringifying it. Rendering a scalar to
-- characters is a host decision, so the interchange format declines to make
-- it.
--
-- 'NElement' holds a @value@ slot alongside its children, for targets that
-- attach a body value to a tagged node (a YAML\/HCL scalar leaf, a config
-- value); it is 'Null' unless a template sets one. Its attributes are an
-- ordered list rather than a map: an element may carry any number of
-- attributes /and/ any number of actions, and source order is preserved.
--
-- 'NFragment' is a real node here -- an ordered sibling sequence with no
-- wrapper element. A host may flatten fragments when folding; evaluation
-- does not.
data Node
  = NText Value Annotations
  | NElement Text [NodeAttribute] Value [Node] Annotations
  | NFragment [Node] Annotations
  deriving stock (Eq, Show)

-- | Both attribute-position constructs, sharing one list. An action's
-- @event@ and @key@ are plain 'Text' because both are static in the source
-- AST (see "Tramaj.Ast"); only the payload is computed.
data NodeAttribute
  = NAttr Text Value
  | NAction Text Text Value
  deriving stock (Eq, Show)

-- Serialization -----------------------------------------------------------

-- | Encodes to the normative representation. Every field is always emitted,
-- including an empty @annotations@ and a @null@ element @value@ -- see
-- @../specs/node-json.md@, "Decoding": on the wire a missing field is a bug,
-- not a default.
nodeToJson :: Node -> Value
nodeToJson (NText v anns) =
  object ["type" .= ("text" :: Text), "value" .= v, "annotations" .= annotationsToJson anns]
nodeToJson (NElement tag attrs val children anns) =
  object
    [ "type" .= ("element" :: Text)
    , "tag" .= tag
    , "attributes" .= V.fromList (map nodeAttributeToJson attrs)
    , "value" .= val
    , "children" .= V.fromList (map nodeToJson children)
    , "annotations" .= annotationsToJson anns
    ]
nodeToJson (NFragment children anns) =
  object
    [ "type" .= ("fragment" :: Text)
    , "children" .= V.fromList (map nodeToJson children)
    , "annotations" .= annotationsToJson anns
    ]

nodeAttributeToJson :: NodeAttribute -> Value
nodeAttributeToJson (NAttr name val) =
  object ["kind" .= ("attribute" :: Text), "name" .= name, "value" .= val]
nodeAttributeToJson (NAction event key payload) =
  object ["kind" .= ("action" :: Text), "event" .= event, "key" .= key, "payload" .= payload]

annotationsToJson :: Annotations -> Value
annotationsToJson = Object . KeyMap.fromList . map (\(k, v) -> (Key.fromText k, v)) . Map.toList

-- | Decodes the normative representation, strictly: a missing or
-- ill-typed field is an error rather than a silently-defaulted value, so
-- @nodeFromJson . nodeToJson@ round-trips and a malformed document is
-- reported where it is read rather than where it later misbehaves. The
-- 'Left' carries a human-readable path-ish description of what was wrong.
nodeFromJson :: Value -> Either String Node
nodeFromJson (Object obj) = do
  ty <- reqString "node" "type" obj
  case ty of
    "text" -> do
      v <- req "text node" "value" obj
      anns <- reqAnnotations obj
      pure (NText v anns)
    "element" -> do
      tag <- reqString "element node" "tag" obj
      attrs <- reqArray "element node" "attributes" obj >>= traverse nodeAttributeFromJson
      val <- req "element node" "value" obj
      children <- reqArray "element node" "children" obj >>= traverse nodeFromJson
      anns <- reqAnnotations obj
      pure (NElement tag attrs val children anns)
    "fragment" -> do
      children <- reqArray "fragment node" "children" obj >>= traverse nodeFromJson
      anns <- reqAnnotations obj
      pure (NFragment children anns)
    other -> Left ("unknown node type: " <> show other)
nodeFromJson _ = Left "expected a JSON object for a node"

nodeAttributeFromJson :: Value -> Either String NodeAttribute
nodeAttributeFromJson (Object obj) = do
  kind <- reqString "node attribute" "kind" obj
  case kind of
    "attribute" -> NAttr <$> reqString "attribute" "name" obj <*> req "attribute" "value" obj
    "action" ->
      NAction
        <$> reqString "action" "event" obj
        <*> reqString "action" "key" obj
        <*> req "action" "payload" obj
    other -> Left ("unknown node attribute kind: " <> show other)
nodeAttributeFromJson _ = Left "expected a JSON object for a node attribute"

req :: String -> Text -> KeyMap.KeyMap Value -> Either String Value
req what field obj = case KeyMap.lookup (Key.fromText field) obj of
  Nothing -> Left (what <> ": missing required field " <> show field)
  Just v -> Right v

reqString :: String -> Text -> KeyMap.KeyMap Value -> Either String Text
reqString what field obj =
  req what field obj >>= \case
    String s -> Right s
    _ -> Left (what <> ": field " <> show field <> " must be a string")

reqArray :: String -> Text -> KeyMap.KeyMap Value -> Either String [Value]
reqArray what field obj =
  req what field obj >>= \case
    Array arr -> Right (V.toList arr)
    _ -> Left (what <> ": field " <> show field <> " must be an array")

reqAnnotations :: KeyMap.KeyMap Value -> Either String Annotations
reqAnnotations obj =
  req "node" "annotations" obj >>= \case
    Object anns -> Right (Map.fromList (map (\(k, v) -> (Key.toText k, v)) (KeyMap.toList anns)))
    _ -> Left "node: field \"annotations\" must be an object"

-- Transformation -----------------------------------------------------------

-- | Rewrites every action reachable in a tree, leaving everything else --
-- tags, ordinary attributes, element value slots, child order, and every
-- annotation map -- untouched. The traversal is effectful in 'Either' so a
-- rewrite that can fail (@adapt-actions@'s optional payload closure) reports
-- where it failed rather than being forced to succeed.
mapActions :: (Text -> Text -> Value -> Either e NodeAttribute) -> Node -> Either e Node
mapActions _ n@(NText _ _) = Right n
mapActions f (NElement tag attrs val children anns) = do
  attrs' <- traverse step attrs
  children' <- traverse (mapActions f) children
  pure (NElement tag attrs' val children' anns)
  where
    step a@(NAttr _ _) = Right a
    step (NAction event key payload) = f event key payload
mapActions f (NFragment children anns) = do
  children' <- traverse (mapActions f) children
  pure (NFragment children' anns)

File: tramaj-rs/src/node.rs
//! `Node`/`NodeAttribute` and the strict `specs/node-json.md` encode/decode
//! (hand-written, not derived, so the "decoder MUST reject ... MUST NOT
//! infer a missing field from a default" rules are enforced exactly).
//! Mirrors `tramaj-hs/src/Tramaj/Node.hs`.

use std::collections::BTreeMap;

use serde_json::{Map as JsonMap, Value as Json};

pub type Annotations = BTreeMap<String, Json>;

pub fn no_annotations() -> Annotations {
    BTreeMap::new()
}

#[derive(Debug, Clone, PartialEq)]
pub enum Node {
    Text(Json, Annotations),
    Element(String, Vec<NodeAttribute>, Json, Vec<Node>, Annotations),
    Fragment(Vec<Node>, Annotations),
}

#[derive(Debug, Clone, PartialEq)]
pub enum NodeAttribute {
    Attribute(String, Json),
    Action(String, String, Json),
}

fn annotations_to_json(anns: &Annotations) -> Json {
    let mut m = JsonMap::new();
    for (k, v) in anns {
        m.insert(k.clone(), v.clone());
    }
    Json::Object(m)
}

pub fn node_to_json(n: &Node) -> Json {
    match n {
        Node::Text(v, anns) => {
            let mut m = JsonMap::new();
            m.insert("type".to_string(), Json::String("text".to_string()));
            m.insert("value".to_string(), v.clone());
            m.insert("annotations".to_string(), annotations_to_json(anns));
            Json::Object(m)
        }
        Node::Element(tag, attrs, val, children, anns) => {
            let mut m = JsonMap::new();
            m.insert("type".to_string(), Json::String("element".to_string()));
            m.insert("tag".to_string(), Json::String(tag.clone()));
            m.insert(
                "attributes".to_string(),
                Json::Array(attrs.iter().map(node_attribute_to_json).collect()),
            );
            m.insert("value".to_string(), val.clone());
            m.insert(
                "children".to_string(),
                Json::Array(children.iter().map(node_to_json).collect()),
            );
            m.insert("annotations".to_string(), annotations_to_json(anns));
            Json::Object(m)
        }
        Node::Fragment(children, anns) => {
            let mut m = JsonMap::new();
            m.insert("type".to_string(), Json::String("fragment".to_string()));
            m.insert(
                "children".to_string(),
                Json::Array(children.iter().map(node_to_json).collect()),
            );
            m.insert("annotations".to_string(), annotations_to_json(anns));
            Json::Object(m)
        }
    }
}

pub fn node_attribute_to_json(a: &NodeAttribute) -> Json {
    match a {
        NodeAttribute::Attribute(name, val) => {
            let mut m = JsonMap::new();
            m.insert("kind".to_string(), Json::String("attribute".to_string()));
            m.insert("name".to_string(), Json::String(name.clone()));
            m.insert("value".to_string(), val.clone());
            Json::Object(m)
        }
        NodeAttribute::Action(event, key, payload) => {
            let mut m = JsonMap::new();
            m.insert("kind".to_string(), Json::String("action".to_string()));
            m.insert("event".to_string(), Json::String(event.clone()));
            m.insert("key".to_string(), Json::String(key.clone()));
            m.insert("payload".to_string(), payload.clone());
            Json::Object(m)
        }
    }
}

fn req<'a>(what: &str, field: &str, obj: &'a JsonMap<String, Json>) -> Result<&'a Json, String> {
    obj.get(field)
        .ok_or_else(|| format!("{what}: missing required field {field:?}"))
}

fn req_string(what: &str, field: &str, obj: &JsonMap<String, Json>) -> Result<String, String> {
    match req(what, field, obj)? {
        Json::String(s) => Ok(s.clone()),
        _ => Err(format!("{what}: field {field:?} must be a string")),
    }
}

fn req_array<'a>(
    what: &str,
    field: &str,
    obj: &'a JsonMap<String, Json>,
) -> Result<&'a Vec<Json>, String> {
    match req(what, field, obj)? {
        Json::Array(a) => Ok(a),
        _ => Err(format!("{what}: field {field:?} must be an array")),
    }
}

fn req_annotations(obj: &JsonMap<String, Json>) -> Result<Annotations, String> {
    match req("node", "annotations", obj)? {
        Json::Object(m) => Ok(m.iter().map(|(k, v)| (k.clone(), v.clone())).collect()),
        _ => Err("node: field \"annotations\" must be an object".to_string()),
    }
}

pub fn node_from_json(v: &Json) -> Result<Node, String> {
    let obj = match v {
        Json::Object(m) => m,
        _ => return Err("expected a JSON object for a node".to_string()),
    };
    let ty = req_string("node", "type", obj)?;
    match ty.as_str() {
        "text" => {
            let value = req("text node", "value", obj)?.clone();
            let anns = req_annotations(obj)?;
            Ok(Node::Text(value, anns))
        }
        "element" => {
            let tag = req_string("element node", "tag", obj)?;
            let attrs = req_array("element node", "attributes", obj)?
                .iter()
                .map(node_attribute_from_json)
                .collect::<Result<Vec<_>, _>>()?;
            let val = req("element node", "value", obj)?.clone();
            let children = req_array("element node", "children", obj)?
                .iter()
                .map(node_from_json)
                .collect::<Result<Vec<_>, _>>()?;
            let anns = req_annotations(obj)?;
            Ok(Node::Element(tag, attrs, val, children, anns))
        }
        "fragment" => {
            let children = req_array("fragment node", "children", obj)?
                .iter()
                .map(node_from_json)
                .collect::<Result<Vec<_>, _>>()?;
            let anns = req_annotations(obj)?;
            Ok(Node::Fragment(children, anns))
        }
        other => Err(format!("unknown node type: {other:?}")),
    }
}

pub fn node_attribute_from_json(v: &Json) -> Result<NodeAttribute, String> {
    let obj = match v {
        Json::Object(m) => m,
        _ => return Err("expected a JSON object for a node attribute".to_string()),
    };
    let kind = req_string("node attribute", "kind", obj)?;
    match kind.as_str() {
        "attribute" => {
            let name = req_string("attribute", "name", obj)?;
            let value = req("attribute", "value", obj)?.clone();
            Ok(NodeAttribute::Attribute(name, value))
        }
        "action" => {
            let event = req_string("action", "event", obj)?;
            let key = req_string("action", "key", obj)?;
            let payload = req("action", "payload", obj)?.clone();
            Ok(NodeAttribute::Action(event, key, payload))
        }
        other => Err(format!("unknown node attribute kind: {other:?}")),
    }
}

/// Rewrites every action reachable in a tree, leaving everything else
/// untouched. Mirrors `Node.hs`'s `mapActions`.
pub fn map_actions<E>(
    n: &Node,
    f: &mut impl FnMut(&str, &str, &Json) -> Result<NodeAttribute, E>,
) -> Result<Node, E> {
    match n {
        Node::Text(v, anns) => Ok(Node::Text(v.clone(), anns.clone())),
        Node::Element(tag, attrs, val, children, anns) => {
            let attrs2 = attrs
                .iter()
                .map(|a| match a {
                    NodeAttribute::Attribute(_, _) => Ok(a.clone()),
                    NodeAttribute::Action(event, key, payload) => f(event, key, payload),
                })
                .collect::<Result<Vec<_>, E>>()?;
            let children2 = children
                .iter()
                .map(|c| map_actions(c, f))
                .collect::<Result<Vec<_>, E>>()?;
            Ok(Node::Element(
                tag.clone(),
                attrs2,
                val.clone(),
                children2,
                anns.clone(),
            ))
        }
        Node::Fragment(children, anns) => {
            let children2 = children
                .iter()
                .map(|c| map_actions(c, f))
                .collect::<Result<Vec<_>, E>>()?;
            Ok(Node::Fragment(children2, anns.clone()))
        }
    }
}
